### PR TITLE
Proper support for allow_partial

### PR DESCRIPTION
**Pull request by EduardoGoulart1**
@eliotwrobson @caleb531 Please do not review it because the code is not ready yet. This is just a draft to book progress on the allow_partial support.

If you go with the definition that DFAs are sets of words, then all operations are well-defined and relatively easy to implement. Most of the times the only required change is to replace loops like:

```Python
for symbol in input_symbols:
    tgt_a = transitions_a[symbol]
```
By the code:
```Python
for symbol in input_symbols:
    tgt_a = transitions_a.get(tgt_a, None)
```

I made the code such that it adds no overhead for complete DFAs compared to the current implementation. For partial DFAs, if the DFAs are sparse, then you will get a large performance boost, while if they are dense you will pay some penalty (but most of the times negligible).

I feel like we still need to discuss a few things:

1. ~~Reserve something (probably None) to represent the trap state. Basically, we add the invariant that if None is part of the set of states, then it *must* be a trap state. This is already implicitly assumed in `_get_next_current_state` and greatly simplifies the implementation logic~~. Because network graph does not allow for states labeled None, we replaced it with automatically generated integers
2. We could think of making this allow_partial thing invisible to the users. Instead of passing allow_partial as a parameter, we compute internally the flag "is_partial". I go by the principle that we should not restrict our API unless it would lead to misuse.
3. Now that support for partial DFAs is there, it is straightforward to extend operations to support DFAs with different alphabets

**Missing:**
- [ ] Adjust/add tests
- [ ] Add benchmarks

### CODE REVIEW

File: automata/fa/dfa.py
Comment (eliotwrobson) [inline automata/fa/dfa.py:241]: Now that I'm thinking of it, the digraph associated with a DFA should probably be made into a cached property.

Comment (EduardoGoulart1) [reply automata/fa/dfa.py:241]: It is funny because I was going to propose quite the opposite. Getting rid of any cached property

Comment (EduardoGoulart1) [reply automata/fa/dfa.py:241]: We don't need to discuss it here, but my problem is specifically with the word cache. Caching all words of length k is a (silent) memory bomb. In my opinion, this should be implemented as an iterator and if the users want fast access to this set, they should compute and store it themselves.

Caching digraph and word lengths seem to offer an acceptable memory overhead (I would still vote for its removal otherwise we need to worry about keeping it consistent)

Comment (caleb531) [reply automata/fa/dfa.py:241]: I do share @EduardoGoulart1's concern about bloating memory usage, but am not sure how to balance it if the alternative comes with a considerable performance impact to many methods. Maybe we make this a global configuration option (i.e. in [`automata.base.config`](https://github.com/caleb531/automata/blob/main/automata/base/config.py)) and leave it up to the user?

I'm a bit undecided on the right path forward, so @eliotwrobson, please let me know your thoughts, and perhaps we can all come to an agreement on the right approach.

Comment (eliotwrobson) [reply automata/fa/dfa.py:241]: I think either a global or local config is the way to go. Kinda like the immutability changes, for most situations, the current memory usage isn't a huge deal, and it makes common operations run quickly (I think that that the accepted words are actually computed using dynamic programming, so memory usage at some level is unavoidable). But I agree that it would be nice to have the option to turn these off if people are running into performance issues. Setting this via global or local config should be fairly straightforward.

I don't think this quite applies yet, but if people are running into more and more performance issues, we could consider introducing Cython as a backend for this project. That will _substantially_ improve things on both of these fronts. However, this would be a significant effort, so maybe something to consider in v9 if this is still a concern.

Comment (caleb531) [reply automata/fa/dfa.py:241]: Okay, great! Since we have a couple different scenarios:

1. Cache words
2. Cache diggraph / word lengths
3. Disabled cache

I wonder if we should have some config property like `cache_mode` or `word_cache_mode` (@eliotwrobson, maybe help me with a good name here 😅), with an enum as the set of possible values.

As for the Cython thing, I have zero experience writing C extensions for Python. It does sound like a very interesting idea though, provided we have someone who knows what they're doing (not sure how familiar you yourself are).

Comment (eliotwrobson) [reply automata/fa/dfa.py:241]: > Okay, great! Since we have a couple different scenarios:
> 
> 1. Cache words
> 2. Cache diggraph / word lengths
> 3. Disabled cache

@caleb531 maybe call this a `cache_level` enum, where higher levels equate to more caching (so cache everything > just cache the digraph > cache nothing). I apologize somewhat for derailing discussion here a bit, it's probably best to handle this after everything else for v8 has already been merged (since the code changes are going to be short and this is a performance improvement and not a new feature).

> As for the Cython thing, I have zero experience writing C extensions for Python. It does sound like a very interesting idea though, provided we have someone who knows what they're doing (not sure how familiar you yourself are).

I'm not really familiar with this either (although I actually think it's fairly straightforward, the code I've seen just looks like Python with some extra information). I think I mentioned this somewhere else, but most of the current implementations have really hit the limit of performance in terms of what can be done in pure Python. Everything here is efficient enough for my purposes, so I think this can wait if/when someone comes along needing to really push the efficiency and wants to take this on.

@EduardoGoulart1 speaking of which, if you find yourself needing minimization for DFAs over large alphabets, there actually is a specialized algorithm that performs better than the one implemented here. So even for this use case, we don't have to resort to introducing new technologies quite yet 😅 .

Comment (EduardoGoulart1) [reply automata/fa/dfa.py:241]: Not sure about how the Cython extensions could look like, but I would prefer to avoid it as much as possible... 
@eliotwrobson which algorithm are you talking about? I only know [this one](https://www.sciencedirect.com/science/article/abs/pii/S0020019011003267)

Comment (EduardoGoulart1) [reply automata/fa/dfa.py:241]: Let's bring this discussion somewhere else, I think that for now we can just go along with this PR (I'll try to wrap it up the next days)

**Comment by EduardoGoulart1** [comment]
I'm currently working on finalizing this PR and I value your opinions before proceeding.

When transitioning from partial to complete DFAs, it is necessary to include a trap state. So far I was using None for this purpose, but it doesn't work well with networkx.DiGraph because that library doesn't allow nodes labeled None in the graph. It is not too difficult to work around this problem for the DFA class, but it becomes problematic when calling NFA.to_dfa (or other automata types) as this invariant would need to propagate throughout the code. So I need another approach to generate new trap states.

The only solution I can think of that does not require significant refactoring, is to assign the highest negative integer number that isn't already a state. This solution works well, but sometimes states are strings or frozen sets, which leads to a typing mismatch that isn't directly caused by the user. To some extent, this can be mitigated on our side, but I don't think that we can completely avoid it.

**Comment by caleb531** [comment]
@EduardoGoulart1 Regarding mixed state types, that's a fair question. @eliotwrobson I think we've updated the library to handle, for example, states of mixed types, correct? I believe some existing methods (maybe `DFA.minify`?) produce machines with such mixed state types.

If this is the case, then I'm fine with the highest negative integer number.

**Comment by eliotwrobson** [comment]
@caleb531 yes, the library can handle mixed state types, and there are a couple of places this is used I think (see `from_finite_language`). @EduardoGoulart1 I think that's a fine convention, as there are a few places where we start assigning integer state names starting from 0 (in the regex code I believe).

**Comment by EduardoGoulart1** [comment]
Also, adapting the test is fairly easy, but some tests have hard-coded values for the expected result which for me does not make sense. Here for instance this test is explicitly testing the union operator by constructing the DFA representing the result.

Is there a specific reason for that? I would have expected the test to verify its algebraic properties. For example something like `A \subseteq (A \cup B) && B \subseteq (A \cub B) && (A \cub B) - A - B == \emptyset`

image

**Comment by eliotwrobson** [comment]
@EduardoGoulart1 the reason for these is to verify the state names and types of the output (which when names are retained, is part of the API). These test cases should be kept with their hard-coded output. There actually is a test case testing for algebraic properties separately.

I'm not sure about the best way to test with allow partial for these. Is there a way to use parameterized test cases (like in pytest) with nose? The DFAs in those test cases are not partial anyway, so I don't think they can really be meaningfully adapted to test for the new behavior in this PR.

**Comment by caleb531** [comment]
@eliotwrobson @EduardoGoulart1 It does look like nose2 supports test parameterization! Although since DFAs can be verbose to construct, I wonder if the decorator signature would get pretty large. But please feel free to play around to find something practical and maintainable:

https://docs.nose2.io/en/latest/params.html

**Comment by EduardoGoulart1** [comment]
@caleb531 @eliotwrobson sure my initial idea was to extend the tests to use dfa.as_partial(). The only places where this does not work is for such hard-coded tests. The problem is that we do not expand all the states. But I will try my best with these constraints in mind :D

**Comment by eliotwrobson** [comment]
@EduardoGoulart1 just a heads up that, because of the size of the refactor and some weirdness that was uncovered, I want to wait until #129 is merged before merging this. I fixed the last major blocker there, so hopefully that will happen soon.

**Comment by caleb531** [comment]
@EduardoGoulart1 @eliotwrobson Just merged #129! I'm working on resolving the merge conflicts now.

File: automata/fa/dfa.py
Comment (eliotwrobson) [inline automata/fa/dfa.py:412]:
```suggestion
        graph.update(self.states)
```
Can this just take in any iterable?

Comment (eliotwrobson) [reply automata/fa/dfa.py]: Updated here https://github.com/caleb531/automata/pull/147/commits/7123f7ec24d8ec0fe1555e1c61aeb213880142d3

**Comment by eliotwrobson** [comment]
@EduardoGoulart1 Some merge conflicts came up when #152 was merged. I've resolved those and gotten the lint check passing in the latest push (with some minor style changes and optimizations). Please pull when you get the chance.

File: automata/fa/dfa.py
Comment (eliotwrobson) [inline automata/fa/dfa.py:1776]: @EduardoGoulart1 This function should have an `as_partial` option. This is because we always know that the state corresponding to the empty set is a trap state, so if `as_partial` is set, we can always omit and avoid traversing the empty set in the BFS. This should substantially speed up this algorithm when used for converting NFAs that come from regexs (since these NFAs have a lot of transitions to the empty set for unused characters).

Comment (EduardoGoulart1) [reply automata/fa/dfa.py:1814]: Ok will do... I am of the more radical opinion that there should only be partial DFAs

Comment (eliotwrobson) [reply automata/fa/dfa.py:1814]: This should be pretty straightforward, as I added some of this functionality in my commits.

Also, I actually think that's not such a radical idea. We can discuss semantics when more things are implemented, but there might be a nice way of making the partial property appear in more places.

**Comment by caleb531** [comment]
@eliotwrobson @EduardoGoulart1 What's outstanding for this ticket before it can be reviewed/merged? I believe this is the last major piece that is necessary for the v8 release.

**Comment by eliotwrobson** [comment]
@caleb531 the changes I made were pretty minor on top of what was here to start. Per previous comments, I think all that's left is to write tests for everything. Not sure how far along @EduardoGoulart1 was with that.

**Comment by EduardoGoulart1** [comment]
Yes, basically the tests... I stopped halfway through them because it's quite a lot to go over :sweat:

**Comment by eliotwrobson** [comment]
> I stopped halfway through them because it's quite a lot to go over

@EduardoGoulart1 could you push what you have (even if partially completed)? This should speed up the review process at least somewhat.

**Comment by caleb531** [comment]
@EduardoGoulart1 Can you please make sure you've pushed up your latest work so @eliotwrobson can finish the tests? I'd love to get this reviewed/merged soon.

File: automata/fa/dfa.py
Comment (eliotwrobson) [inline automata/fa/dfa.py:560]: @EduardoGoulart1 Why make this change? Doesn't seem necessary.

**Comment by eliotwrobson** [comment]
@EduardoGoulart1 @caleb531 in the latest push, I streamlined and update the test cases to incorporate partial DFAs, including tests that check for interoperability with partial and non-partial DFAs. This bought the line coverage up to 100% in the DFA file.

There are a couple of minor things left to do. The main thing is double checking over the test suite to ensure we are testing extensively enough for the partial behavior. The other thing is doing a pass over the DFA code to clean things up as much as possible. There were places where I removed a large amount of code, but I may have missed some things.

Even though I'm not quite ready to call this done, I think this is in a mature enough state that we can start reviewing, as it should be mostly code complete at this point.

File: automata/fa/dfa.py
Comment (EduardoGoulart1) [reply automata/fa/dfa.py:563]: If the DFA is partial, transitions[state][letter] might not be defined. That's why I went with this one...

**Comment by EduardoGoulart1** [comment]
> @EduardoGoulart1 @caleb531 in the latest push, I streamlined and update the test cases to incorporate partial DFAs, including tests that check for interoperability with partial and non-partial DFAs. This bought the line coverage up to 100% in the DFA file.
> 
> There are a couple of minor things left to do. The main thing is double checking over the test suite to ensure we are testing extensively enough for the partial behavior. The other thing is doing a pass over the DFA code to clean things up as much as possible. There were places where I removed a large amount of code, but I may have missed some things. Finally, we need to update the documentation.
> 
> Even though I'm not quite ready to call this done, I think this is in a mature enough state that we can start reviewing, as it should be mostly code complete at this point.

Does it make sense to start over. Now I am a bit lost in the middle of all the recent updates

**Comment by eliotwrobson** [comment]
@EduardoGoulart1 I don't think it makes sense to make a new one, but you'll probably have to review the entire current set of changes instead of any one commit.

**Comment by EduardoGoulart1** [comment]
Just had a quick look over it. I would say it's in a state to start reviewing it, so I will just drop my review...

**Comment by eliotwrobson** [comment]
> But who is in charge of applying the changes?

Depending on the number of changes, I can probably apply them

File: tests/test_dfa.py
Comment (EduardoGoulart1) [inline tests/test_dfa.py:291]: This test should test the partial DFAs (applies to some other tests too). I was doing something like this:

```
def partial_combinations(*args):
    args = tuple(dfa.copy() for dfa in args)
    for partial_flags in product((False, True), repeat=len(args)):
        yield partial_flags, (
            to_partial(dfa, partial_flag)
            for partial_flag, dfa in zip(partial_flags, args)
        )
```

Comment (EduardoGoulart1) [reply tests/test_dfa.py:291]: Then for each test I would just pass the DFAs to the `partial_combinations` function and apply the asserts normally

File: tests/test_dfa.py
Comment (EduardoGoulart1) [inline tests/test_dfa.py:740]: I think that here I prefer the original names... Makes it more understandable

**Review by EduardoGoulart1 — COMMENTED** [review]
I am not so sure about the changes to the BFS method, otherwise minor changes

I think that the tests can be improved to try all combinations of the partial/complete DFAs with the code that I sent, but to be honest at this point I think that it's easier to just bring this PR to completion, so what we have is already good enough

File: automata/fa/dfa.py
Comment (eliotwrobson) [reply automata/fa/dfa.py:863]: I think the fix for this is actually pretty straightforward, just iterate through all input symbols in both inner dictionaries instead of the input symbol set each time in the inner loop.

Comment (EduardoGoulart1) [reply automata/fa/dfa.py:863]: This is already good enough, but would till be a problem if one of the DFAs is dense

File: automata/fa/nfa.py
Comment (eliotwrobson) [reply automata/fa/nfa.py:223]: Would normally agree but the return type of this function is to match a specific signature somewhere else. Take a look at the from_nfa function.

File: automata/fa/dfa.py
Comment (EduardoGoulart1) [reply automata/fa/dfa.py:863]: Also minor comment: There are some TODOs introduced, I think they can be removed or we create an issue for them

File: automata/fa/dfa.py
Comment (eliotwrobson) [reply automata/fa/dfa.py:1362]: Cheap way of saving memory since these are all integers.

**Comment by EduardoGoulart1** [comment]
Since there are only minor comments from my side, and I don't want to block progress because of my limited time, I am good with whatever decisions you meet here so we can merge it (I can't approve my own PR)

File: tests/test_dfa.py
Comment (eliotwrobson) [reply tests/test_dfa.py:740]: The original names are no longer descriptive. If you take a look at the parametrization of this function, it now tries all pairs of distinct DFAs from the parameters list instead of just two example DFAs.

This is actually the main test case checking for valid interop with complete and partial DFAs now.

File: automata/fa/dfa.py
Comment (eliotwrobson) [reply automata/fa/dfa.py:313]: I'll double check, but I think the old code always added a trap state. This function is for internal use only, and may get inputs that are already complete DFAs. Basically, the change in code is accompanying a change in the use case.

File: automata/fa/dfa.py
Comment (eliotwrobson) [reply automata/fa/dfa.py:863]: I think that efficiency issue is very specific to the intersection with dense DFAs. I'll leave a Todo that this can possibly be improved by writing a helper function to case on the booleans used as inputs, and we can implement this if anyone ever comes along with this issue.

As for the todos, they were mainly for myself as I didn't write tests for a couple of edge cases with partial DFAs. I'll remove those and add tests later today at some point.

File: automata/fa/dfa.py
Comment (eliotwrobson) [reply automata/fa/dfa.py:1210]: It could but it seems unnecessary, the only speed increase would come from removing dict lookups, and I'm not sure there's any real benefit to that.

File: automata/fa/dfa.py
Comment (caleb531) [reply automata/fa/dfa.py:1362]: @eliotwrobson I'm fine with that, but can you please add a code comment explaining why `array.array` is used here?

File: tests/test_dfa.py
Comment (caleb531) [reply tests/test_dfa.py:740]: @eliotwrobson I think the new names are fine, and agree with @eliotwrobson's reasoning for something more descriptive. Closing this thread.

File: automata/fa/dfa.py
Comment (caleb531) [reply automata/fa/dfa.py:563]: @EduardoGoulart1 Makes sense to be. Let's keep it the way it is.

File: automata/fa/nfa.py
Comment (caleb531) [reply automata/fa/nfa.py:223]: @EduardoGoulart1 Agreed with @eliotwrobson. `_iterate_through_symbol_path_pairs` needs to be a generator, meaning we need to `yield` the individual values instead of returning the entire iterable.

**Comment by caleb531** [comment]
I've worked through a few of the comment threads, closing the ones that are taken care of, and commenting on a few others.

@EduardoGoulart1 @eliotwrobson Would it be possible for you guys to work through the outstanding comment threads? I would personally prefer to review this PR only once these comment threads have been resolved, and I don't have anything more to say on the ones that are still open.

**Comment by eliotwrobson** [comment]
> Would it be possible for you guys to work through the outstanding comment threads? I would personally prefer to review this PR only once these comment threads have been resolved, and I don't have anything more to say on the ones that are still open.

@caleb531 sounds good, I'm resolving some now and the remaining threads are fairly short.

File: automata/fa/dfa.py
Comment (eliotwrobson) [reply automata/fa/dfa.py:290]: Added check here with test in a later commit: https://github.com/caleb531/automata/pull/147/commits/85d8be7b4c8d584eb2f548062b3736959495381e

File: automata/fa/dfa.py
Comment (eliotwrobson) [reply automata/fa/dfa.py:319]: Use case changed slightly so I was able to switch back to the old code: https://github.com/caleb531/automata/pull/147/commits/b7fa44db7a80ab727d4324b389e3f8966e708c7e

File: automata/fa/dfa.py
Comment (eliotwrobson) [reply automata/fa/dfa.py:1375]: Comment added here: https://github.com/caleb531/automata/pull/147/commits/b7fa44db7a80ab727d4324b389e3f8966e708c7e

File: automata/fa/dfa.py
Comment (eliotwrobson) [reply automata/fa/dfa.py:869]: Simplified logic and added comment per discussion: https://github.com/caleb531/automata/pull/147/commits/85d8be7b4c8d584eb2f548062b3736959495381e

File: automata/fa/fa.py
Comment (eliotwrobson) [reply automata/fa/fa.py]: Removed here https://github.com/caleb531/automata/pull/147/commits/2a72416dc1a16d120c9be0c4e718ca843865003e

File: tests/test_dfa.py
Comment (eliotwrobson) [reply tests/test_dfa.py:296]: This is subsumed by the `test_set_laws` function, where the equality function is used on a mix of partial and complete DFAs.

File: automata/fa/dfa.py
Comment (eliotwrobson) [reply automata/fa/dfa.py:542]: Sorry, I was a bit unclear in my original comment. It makes sense why this line was changed, but we can call `keys()` instead of `items()` since the `tgt_state` variable isn't being read. I'll update the code with this change.

**Comment by eliotwrobson** [comment]
@caleb531 all items are resolved and removed all TODOs. The only thing left is docs changes, but I'll leave that until after your review. Thanks!

File: automata/fa/dfa.py
Comment (caleb531) [inline automata/fa/dfa.py:48]: @eliotwrobson @EduardoGoulart1 I believe `BooleanLiteral` is unnecessary. You can simply use `bool` as the type hint (e.g. `lhs_relevant: bool` instead of `lhs_relevant: BooleanLiteral`).

File: automata/fa/dfa.py
Comment (caleb531) [inline automata/fa/dfa.py:548]: @EduardoGoulart1 @eliotwrobson I think a DFA can also be considered partial if the top-level transitions dict doesn't include all of the states as its keys (i.e. any states not listed are trap states). For example:

```python
# A DFA whose language is just {"a"}
dfa = DFA(
    states={"q0", "q1"},
    input_symbols={"a", "b"},
    transitions={
        # This case is already covered in the above check, since the transition to "b" is left out
        "q0": {"a": "q1"}
        # Definition for "q1" is omitted because it has no transitions anyway;
        # this case is not considered in the above check to determine a DFA"s partiality
    },
    initial_state="q0",
    final_states={"q1"},
    allow_partial=True,
)
```

Please correct me if I misunderstand the code here, but it just caught my eye.

File: automata/fa/dfa.py
Comment (caleb531) [inline automata/fa/dfa.py:834]: @EduardoGoulart1 @eliotwrobson For clarity, I request that the two new parameters at the end be keyword-only. And, as I mentioned in my above comment, you should be able to just use `bool` for the type hints. That is:

```python
def _cross_product(lhs: DFA, rhs: DFA, *, lhs_relevant: bool, rhs_relevant: bool) -> Tuple[DFAStateT, ExpandStateFn]:
    # ...
```

Even though the method is internal, this will make it much clearer later in the code where you are currently passing `True, True` to this method.

File: automata/fa/dfa.py
Comment (caleb531) [inline automata/fa/dfa.py:229]: @EduardoGoulart1 @eliotwrobson This new method needs to be added to the API Documentation (in `docs/fa/class-dfa.md`).

**Review by caleb531 — CHANGES_REQUESTED** [review]
@EduardoGoulart1 @eliotwrobson Left a few requested changes. Nothing major (at least from my perspective 😅).

File: automata/fa/dfa.py
Comment (eliotwrobson) [reply automata/fa/dfa.py]: @caleb531 agreed, done here! https://github.com/caleb531/automata/pull/147/commits/070ff3206873feee4728c796792b87ff9d88914f

File: automata/fa/dfa.py
Comment (eliotwrobson) [reply automata/fa/dfa.py:546]: @caleb531 this functionality was removed here, take a look at the commit message: https://github.com/EduardoGoulart1/automata/commit/43f377c4b7aa6381c286679a0379776857e2459c

I don't think it's worth adding support for this, since it would lead to some incompatibilities with existing code, and we'd have to go back and test some extra things. This only leads to minor memory savings in very limited circumstances anyway.

Comment (caleb531) [reply automata/fa/dfa.py:546]: @eliotwrobson Okay, fair enough. The current API documentation doesn't say anything against your recommended behavior anyway

> `allow_partial`: by default, each DFA state must have a transition to
every input symbol; if `allow_partial` is `True`, you can disable this
characteristic (such that any DFA state can have fewer transitions than input
symbols)

File: automata/fa/dfa.py
Comment (eliotwrobson) [reply automata/fa/dfa.py:227]: @caleb531 updated here!

Comment (caleb531) [reply automata/fa/dfa.py:227]: @eliotwrobson Yep, just reviewed it! Looks great to me!

**Review by caleb531 — APPROVED** [review]

**Review by eliotwrobson — APPROVED** [review]
@caleb531 Resolved all threads from your comments! If those were all the items you had questions about, feel free to merge when ready!

**Comment by caleb531** [comment]
@EduardoGoulart1 @eliotwrobson This looks good to me! Will merge now.

**Pull request merged**
